File: src/54.spiral-matrix.java
/*
 * @lc app=leetcode id=54 lang=java
 *
 * [54] Spiral Matrix
 */

// @lc code=start
import java.util.*;
class Solution {
    /*
    track当前遍历的row,col范围
    注意：半边之后要重新判断防止重复遍历，eg: [3 1],只需要遍历半边就可以了
    time: O(mn)
    space: O(mn)
    */
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        if (matrix == null || matrix.length == 0) return res;
        int m = matrix.length, n = matrix[0].length;
        int rowBegin = 0, colBegin = 0;
        int rowEnd = m - 1, colEnd = n - 1;
        while (rowBegin <= rowEnd && colBegin <= colEnd) {
            for (int i = colBegin; i <= colEnd; ++i) {
                res.add(matrix[rowBegin][i]);
            }
            rowBegin++;
            for (int i = rowBegin; i <= rowEnd; ++i) {
                res.add(matrix[i][colEnd]);
            }
            colEnd--;
            if (rowBegin <= rowEnd) {
                for (int i = colEnd; i>= colBegin; --i) {
                    res.add(matrix[rowEnd][i]);
                }
                rowEnd--;
            }
            if (colBegin <= colEnd) {
                for (int i = rowEnd; i>= rowBegin; --i) {
                    res.add(matrix[i][colBegin]);
                }
                colBegin++;
            }
        }
        return res;
    }
}
// @lc code=end

File: Cpp/54.spiral-matrix.cpp
/*
 * @lc app=leetcode id=54 lang=cpp
 *
 * [54] Spiral Matrix
 */

// @lc code=start
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        if (matrix.empty()) return res;
        int rowMin = 0, rowMax = matrix.size() - 1, colMin = 0, colMax = matrix[0].size() - 1;
        while (rowMin <= rowMax && colMin <= colMax) {
            for (int i = colMin; i <= colMax; ++i) {
                res.push_back(matrix[rowMin][i]);
            }
            rowMin++;
            for (int i = rowMin; i <= rowMax; ++i) {
                res.push_back(matrix[i][colMax]);
            }
            colMax--;
            if (rowMin <= rowMax) {
                for (int i = colMax; i >= colMin; --i) {
                    res.push_back(matrix[rowMax][i]);
                }
                rowMax--;
            }
            if (colMin <= colMax) {
                for(int i = rowMax; i >= rowMin; --i) {
                    res.push_back(matrix[i][colMin]);
                }
                colMin++;
            }
        }
        return res;
    }
};
// @lc code=end

